oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/FetchJwksHandler.java

```diff
@@ -0,0 +1,60 @@
+package uk.gov.di.authentication.oidc.lambda;
+
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import com.nimbusds.jose.KeySourceException;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKMatcher;
+import com.nimbusds.jose.jwk.JWKSelector;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.jwk.source.JWKSourceBuilder;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.oauth2.sdk.OAuth2Error;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import uk.gov.di.authentication.oidc.entity.FetchJwksResponse;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+public class FetchJwksHandler implements RequestHandler<Map<String, String>, FetchJwksResponse> {
+    FetchJwksResponse response = new FetchJwksResponse();
+    private static final Logger LOG = LogManager.getLogger(FetchJwksHandler.class);
+
+    @Override
+    public FetchJwksResponse handleRequest(Map<String, String> event, Context context) {
+        String url = event.get("url");
+        String keyId = event.get("keyId");
+        if (url == null || keyId == null) {
+            throw new IllegalArgumentException("FetchJwksHandler invoked with invalid argument(s)");
+        }
+        try {
+            JWKSelector selector = new JWKSelector(new JWKMatcher.Builder().keyID(keyId).build());
+            JWKSource<SecurityContext> jwkSource =
+                    JWKSourceBuilder.create(new URL(url))
+                            .retrying(true)
+                            .refreshAheadCache(false)
+                            .cache(false)
+                            .rateLimited(false)
+                            .build();
+            LOG.info("Fetching JWKS from URL: {}", url);
+            JWK jwk =
+                    jwkSource.get(selector, null).stream()
+                            .findFirst()
+                            .orElseThrow(
+                                    () -> new KeySourceException("No key found with given keyID"));
+            response.setJwk(jwk);
+        } catch (KeySourceException e) {
+            LOG.error("Failed to fetch JWKS: could not find key with provided key ID", e);
+            response.setError(OAuth2Error.SERVER_ERROR);
+        } catch (MalformedURLException e) {
+            LOG.error("Failed to fetch JWKS: URL is malformed", e);
+            response.setError(OAuth2Error.SERVER_ERROR);
+        } catch (IllegalArgumentException e) {
+            LOG.error("Failed to fetch JWKS: url and/or keyId parameters not present", e);
+            response.setError(OAuth2Error.SERVER_ERROR);
```

Comment:
Is an OAuth2Error appropriate here? If so, is SERVER_ERROR reasonable?

Reply:
It feels a bit odd to be using this here. I agree this is what we should be returning to the RP, but it feels like we're leaking that context into a function that could be pretty separate from that

Reply:
Have updated it, see what you think